### PR TITLE
feat(compiler): Implement constant string concat propagation

### LINKED ISSUE #29617: Compiler String Concat Constant Propagation
Consider the following constant propagation:

Input:
```ts
function foo() {
  const a = "a" + "b";
  const c = "c";
  return a + c;
}
```
Output:
```ts
function foo() {
  return "abc";
}
```

Is this in scope for the react compiler? If so, is it open for contributions?

### PR DESCRIPTION
**Pull request by nikeee**
## Summary

Resolves #29617

## How did you test this change?
I verified the implementation using the test.

### CODE REVIEW

File: compiler/crates/react_hermes_parser/tests/fixtures/constant-propagation-string-concat.js
Comment (josephsavona) [inline compiler/crates/react_hermes_parser/tests/fixtures/constant-propagation-string-concat.js:1]: the rust port of the compiler is unused and unmaintained and we aren't accepting contributions to this part, can you revert the changes here? thanks!

**Review by josephsavona — COMMENTED** [review]
The changes look good, let's just revert the changes under compiler/crates/

File: compiler/crates/react_hermes_parser/tests/fixtures/constant-propagation-string-concat.js
Comment (nikeee) [reply compiler/crates/react_hermes_parser/tests/fixtures/constant-propagation-string-concat.js:1]: Done

**Comment by nikeee** [comment]
I was wondering about the cases where one of the operands is string and the other is number: `"a" + 1` and `1 + "0"`. Should we also consider this as in-scope? If yes, we could change the propagation to just be 
```js
case "+": {
  if ((typeof lhs === "number" || typeof lhs === "string")) && (typeof rhs === "number" ||  typeof rhs === "string")) {
    result = { kind: "Primitive", value: lhs + rhs, loc: value.loc };
  }
  break;
}
```

**Comment by josephsavona** [comment]
We'd have to double-check the spec for "+" with string and numbers. Let's start with just number+number and string+string.

**Review by josephsavona — APPROVED** [review]

**Comment by zhangenming** [comment]
What about string+number or num+string
What about other operations(- * / ...)
